atlasdb-autobatch/src/main/java/com/palantir/atlasdb/autobatch/Autobatchers.java

```diff
@@ -79,25 +94,89 @@ private Autobatchers() {}
      * the result for that input element
      * @param <I> type of input element
      * @param <O> type of output element
-     * @return builder where the autobatch can be further customised
+     * @return builder where the autobatcher can be further customised
      */
     public static <I, O> AutobatcherBuilder<I, O> independent(Consumer<List<BatchElement<I, O>>> batchFunction) {
-        return new AutobatcherBuilder<>(bufferSize -> new IndependentBatchingEventHandler<>(batchFunction, bufferSize));
+        return new AutobatcherBuilder<>(parameters -> new IndependentBatchingEventHandler<>(
+                maybeWrapWithTimeout(batchFunction, parameters), parameters.batchSize()));
+    }
+
+    private static <I, O> Consumer<List<BatchElement<I, O>>> maybeWrapWithTimeout(
+            Consumer<List<BatchElement<I, O>>> batchFunction, EventHandlerParameters parameters) {
+        return parameters
+                .batchFunctionTimeoutContext()
+                .map(context -> wrapWithTimeout(batchFunction, parameters.safeLoggablePurpose(), context))
+                .orElse(batchFunction);
+    }
+
+    private static <I, O> CoalescingRequestFunction<I, O> maybeWrapWithTimeout(
+            CoalescingRequestFunction<I, O> coalescingFunction, EventHandlerParameters parameters) {
+        return parameters
+                .batchFunctionTimeoutContext()
+                .map(context -> wrapWithTimeout(coalescingFunction, parameters.safeLoggablePurpose(), context))
+                .orElse(coalescingFunction);
+    }
+
+    private static <I, O> Consumer<List<BatchElement<I, O>>> wrapWithTimeout(
+            Consumer<List<BatchElement<I, O>>> delegate,
+            String safeLoggablePurpose,
+            TimeoutOrchestrationContext context) {
+        TimeLimiter limiter = SimpleTimeLimiter.create(context.exclusiveExecutor());
+        return elements -> {
+            try {
+                limiter.runWithTimeout(() -> delegate.accept(elements), context.batchFunctionTimeout());
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+                throw new RuntimeException(e);
+            } catch (TimeoutException e) {
+                log.info(
+                        "Autobatcher timed out and request had to be interrupted",
+                        SafeArg.of("safeLoggablePurpose", safeLoggablePurpose),
+                        e);
+                throw new RuntimeException(e);
+            } catch (UncheckedExecutionException e) {
+                Throwables.throwIfUnchecked(e.getCause());
+                throw new RuntimeException(e.getCause());
```

Comment:
This seems like something that would be worth having in our `Throwables` class

Reply:
will add